### PR TITLE
Fix failure when pruning partition columns with a date transform applied in Iceberg connector

### PR DESCRIPTION
**Pull request by chenjian2664**
Recreated for https://github.com/trinodb/trino/pull/8101.

Fixes: #7502 
Fixes: #7999

### CODE REVIEW

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsMaker.java [hunk 1]
```diff
@@ -244,6 +241,29 @@ private NullableValue makeNullableValue(io.trino.spi.type.Type type, Object valu
...
+    }
+
+    private void buildIdPrimitiveTypeMapping(Types.NestedField field, ImmutableMap.Builder<Integer, Type.PrimitiveType> icebergIdToTypeMapping)
+    {
+        Type type = field.type();
+        if (type.isPrimitiveType()) {
+            icebergIdToTypeMapping.put(field.fieldId(), type.asPrimitiveType());
+            return;
+        }
+        if (type instanceof Type.NestedType) {
```
Comment (alexjo2144) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsMaker.java:260]: Sorry, I'm not seeing how these changes are related. This has to do with statistics for ROW type sub-columns?

Comment (alexjo2144) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsMaker.java:260]: Ah, I see, there were two issues. Can you add the example from https://github.com/trinodb/trino/issues/7502 to the tests?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/PartitionTable.java [hunk 2]
```diff
@@ -291,6 +291,17 @@ public static Object convert(Object value, Type type)
...
         }
+        if (type instanceof Types.IntegerType || type instanceof Types.DateType) {
+            // date transform applied
+            if (value instanceof Integer) {
+                int actual = (int) value;
+                return (long) actual;
+            }
+            if (value instanceof Long) {
+                return value;
+            }
```
Comment (alexjo2144) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/PartitionTable.java:302]: Nit: maybe just:
```suggestion
            if (value instanceof Number) {
                return ((Number) value).longValue();
            }
```

Comment (hashhar) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/PartitionTable.java:302]: Please don't use cast to Number. It hides a lot of possible bugs.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/PartitionTable.java [hunk 3]
```diff
@@ -291,6 +291,17 @@ public static Object convert(Object value, Type type)
             // TODO the client sees the bytearray's tostring ouput instead of seeing actual bytes, needs to be fixed.
             return ((ByteBuffer) value).array();
         }
+        if (type instanceof Types.IntegerType || type instanceof Types.DateType) {
```
Comment (hashhar) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/PartitionTable.java:294]: You can break both cases into their own branches since the `instanceof Long` should not be happening for Integer type. The current code feels a bit misleading.

Same for DateType - the underlying value is either going to be Integer or Long - it cannot be both.

Any cases where the alternative branch gets taken?

**Comment by hashhar** [comment]
Also it would nice to add a test that directly excercises statistics which is the cause of the issue as we discussed in https://github.com/trinodb/trino/pull/8101#discussion_r649686857.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/AbstractTestIcebergConnectorTest.java [hunk 4]
```diff
@@ -1835,6 +1835,116 @@ public void testIncorrectIcebergFileSizes()
...
+
+        String expectedValues1 = "VALUES (1,'b'), (2, 'c'), (3, 'd'),(4, NULL)";
+        String expectedvalues2 = "VALUES " +
+                "(1, 'b'), " +
+                "(2, 'c'), " +
+                "(3, 'd')," +
+                "(4, NULL), " +
+                "(5, 'f') ";
+
+        // partitioned tables
```
Comment (losipiuk) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/AbstractTestIcebergConnectorTest.java]: please initialize the tables with data in this section. And do only assertions in following sections.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/PartitionTable.java (hunk 3 above)
Comment (losipiuk) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/PartitionTable.java:294]: +1

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/AbstractTestIcebergConnectorTest.java [hunk 5]
```diff
@@ -1836,113 +1836,97 @@ public void testIncorrectIcebergFileSizes()
...
-        assertQuery("SELECT count(pt4.a) FROM test_join_partitioning1 pt1 JOIN test_join_partitioning4 pt4 ON pt1.a = pt4.a", "VALUES 6");
-        assertQuery("SELECT count(pt4.b) FROM test_join_partitioning1 pt1 JOIN test_join_partitioning4 pt4 ON pt1.a = pt4.a", "VALUES 5");
-
-        dropTable("test_join_partitioning1");
-        dropTable("test_join_partitioning2");
-        dropTable("test_join_partitioning3");
-        dropTable("test_join_partitioning4");
-        dropTable("test_join_partitioning_normal1");
-        dropTable("test_join_partitioning_normal2");
+    public void testPartitionedTableWithDateTransformStatistics()
```
Comment (losipiuk) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/AbstractTestIcebergConnectorTest.java:1839]: I think you did not mean to remove this test you have just added. It seems like you are fixing two separate things. It would be best if those were separate PRs.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/PartitionTable.java [hunk 6]
```diff
@@ -300,7 +300,7 @@ public static Object convert(Object value, Type type)
             if (value instanceof Long) {
                 return value;
             }
-            throw new IllegalArgumentException("Unexpected type : " + type);
+            throw new IllegalArgumentException("Unexpected type of value : " + value);
```
Comment (losipiuk) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/PartitionTable.java:303]: More accurate commit message would be `Fix statistics computation for nested types`

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsMaker.java [hunk 7]
```diff
@@ -244,6 +241,29 @@ private NullableValue makeNullableValue(io.trino.spi.type.Type type, Object valu
         return partitionTypeBuilder.build();
     }
 
+    private Map<Integer, Type.PrimitiveType> buildIdPrimitiveTypeMapping(List<Types.NestedField> columns)
```
Comment (losipiuk) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsMaker.java:244]: add "To" to method name `buildIdToPrimitiveTypeMapping`

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsMaker.java [hunk 8]
```diff
@@ -244,6 +241,29 @@ private NullableValue makeNullableValue(io.trino.spi.type.Type type, Object valu
...
+    {
+        ImmutableMap.Builder<Integer, Type.PrimitiveType> idToTypeMapping = ImmutableMap.builder();
+        for (Types.NestedField field : columns) {
+            buildIdPrimitiveTypeMapping(field, idToTypeMapping);
+        }
+        return idToTypeMapping.build();
+    }
+
+    private void buildIdPrimitiveTypeMapping(Types.NestedField field, ImmutableMap.Builder<Integer, Type.PrimitiveType> icebergIdToTypeMapping)
```
Comment (losipiuk) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsMaker.java:253]: add "To" to method name

Comment (losipiuk) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsMaker.java:253]: rename `icebergIdToTypeMapping` to `mappingBuilder`

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsMaker.java (hunk 1 above)
Comment (losipiuk) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsMaker.java:260]: I do not understand:
* which bug this commit is supposed to fix
* how the fix here relate to added test. The test does not seem to use nested types.

Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsMaker.java:260]: > * how the fix here relate to added test.

indeed, the `testPartitionedTableWithDateTransformStatistics` test passes when run without other changes.
